mlir/lib/IR/AffineExpr.cpp

```diff
@@ -406,12 +419,25 @@ static bool isDivisibleBySymbol(AffineExpr expr, unsigned symbolPos,
     AffineBinaryOpExpr binaryExpr = cast<AffineBinaryOpExpr>(expr);
     if (opKind != expr.getKind())
       return false;
-    return isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, expr.getKind());
+    if (llvm::any_of(visitedExprs, [](auto expr) {
```

Comment:
I don't understand this condition. I didn't understand the condition before either, but I still don't understand it. Won't this say `that s0 // 42 + s0` is divisible by `s0`? 

I think we should have unit tests for isDivisibleBySymbol, not just integration-style tests for all of simplification.

Reply:
`a / b / s0 <=> a / s0 / b`. If `a` is a multiple of `s0`, we can simplify a. but if (a / b) * d / s0 does not hold.

Reply:
Ah, so the function is just named incorrectly. How about `canSinkDivisionBySymbol` or something along those lines?

Reply:
Let's do this rename too. `canSimplifyDivisionBySymbol` or `canSinkDivisionBySymbol` would work for me. Maybe the former is better.

Reply:
renamed, sorry my written English is not good.